spec/Classes.tex

```diff
@@ -663,73 +604,456 @@ \subsection{User-Defined Constructors}
+\end{chapelexample}
+
+\subsubsection{Initializing Fields in Conditional Statements}
+\label{Initializing_Fields_in_Conditional_Statements}
+\index{classes!initializers!user-defined!conditionals}
+
+Fields may be initialized inside of conditional statements, with the
+restriction that the same set of fields must be initialized in every branch.
+If the user omits any field initializations, the compiler will insert field
+initializations up to and including the field furthest in field declaration
+order between the conditional branches. If the else branch of a conditional
+statement is omitted, the compiler will generate an empty else branch and
+insert field initialization statements as needed.
+
+\begin{chapelexample}{initFieldConditional.chpl}
+In the following code:
+\begin{chapel}
+class Point {
+  var x, y : real;
+  var r, theta : real;
+
+  proc init(polar : bool, val : real) {
+    if polar {
+      // compiler inserts initialization for fields 'x' and 'y'
+      this.r = val;
+    } else {
+      this.x = val;
+      this.y = val;
+      // compiler inserts initialization for field 'r'
+    }
+    // compiler inserts initialization for field 'theta'
+  }
+}
+
+var A = new Point(true, 5.0);
+var B = new Point(false, 1.0);
+\end{chapel}
+\begin{chapelpost}
+writeln(A);
+writeln(B);
+\end{chapelpost}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+{x = 0.0, y = 0.0, r = 5.0, theta = 0.0}
+{x = 1.0, y = 1.0, r = 0.0, theta = 0.0}
+\end{chapeloutput}
+
+The compiler identifies field \chpl{r} as the latest field in both branches,
+and inserts omitted field initialization statements as needed to ensure that
+fields \chpl{x}, \chpl{y}, and \chpl{r} are all initialized by the end of the
+conditional.
+\end{chapelexample}
+
+Conditionals may also contain calls to sibling and parent initializers,
+provided that the initialization state is the same at the end of the
+conditional statement.
+
+\begin{chapelexample}{thisDotInitConditional.chpl}
+In the following code:
+\begin{chapel}
+class Parent {
+  var x, y : real;
+}
+
+class Child : Parent {
+  var z : real;
+
+  proc init(cond : bool, val : real) {
+    if cond {
+      super.init(val, val);
+      this.z = val;
+      this.complete();
+    } else {
+      this.init(val, val, val);
+    }
+  }
+
+  proc init(x: real, y: real, z: real) {
+    super.init(x, y);
+    this.z = z;
+  }
+}
+
+var c = new Child(true, 5.0);
+\end{chapel}
+\begin{chapelpost}
+writeln(c);
+\end{chapelpost}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+{x = 5.0, y = 5.0, z = 5.0}
+\end{chapeloutput}
+The first initializer must invoke the \chpl{complete} method at the end of
+the if-branch in order to match the state at the end of the else-branch.
+\end{chapelexample}
+
+\subsubsection{Miscellaneous Field Initialization Rules}
+
+Fields may not be initialized within loop statements, parallel statements, or
+on statements.
```

Comment:
Fields can be initialized in `on` statements.  It looks like the comment in test/classes/initializers/phase1/hasOnFieldInit.chpl is out of date (sorry, that's probably my fault)